Implementation of Neo4jIgnoreAttribute to control property serialization instead of JsonIgnoreAttribute

Clooney24 opened the pull request:
Implementation of Neo4jIgnoreAttribute as BoltGraphClient serializes object parameters ignoring JsonContractResolver, see #336 
Works with BoltGraphClient and GraphClient

Clooney24 commented:
Changes in Neo4jClient/Neo4jClient.csproj and Neo4jClient/project.json were just to get it built on my system.

cskardon commented:
Hey,

I've gone through and would have no problem pulling this in, but I'd need there to be some tests - i.e. proving that it works correctly, especially given that this would replace a default in the client (the contract resolver) from being an imported one to a custom one. 

So tests about Ignoring and renaming would be great!

Clooney24 commented:
Perfect, unfortunally, I have no resources to do more tests. As the Neo4jContractResolver part is just for the non-bolt client, we do not need it for our actual problem switching to BoltGraphClient. I thought it would be easier for documentation when Neo4jIgnoreAttribute works in both clients by default. Maybe you can just include the Neo4jContractResolver in the project but not set it as default, so users can refer to it just by need. We were using it since about 2 years now in our application without problems, so it works fine for us, but I have no idea which different scenarios could be affected on other users applications. If you could just pull in the changes in SerializeObject and the Neo4jIgnoreAttribute it would be perfect.

cskardon commented:
I mean - I'd still need tests to prove that it works - I can't just pull in code without proof - from my pov - if I put in the `Neo4jIgnoreAttribute` and someone uses it and it doesn't work - it's on me to fix it then - it's especially an issue when it's around the serialization aspect.

I can see the benefit of the code, and I'd hope to get time to add tests myself. Do I have permission to push to your branch?

Clooney24 commented:
What kind of proof do you need exactly? Before I created the PR, I tested it in a branch of our full recruiting portal web application and everything worked fine (I removed the official Neo4jClient-Nuget package, included the project from the PR and switched it from https to bolt - there is heavy use of [Neo4jIgnore] in our application and it didn't worked over bolt with the Nuget version). 

If you don't make the Neo4jContractResolver as default, it's just this one line which has any influence on behaviour and could break:  ```pi.IsDefined(typeof(JsonIgnoreAttribute)) || pi.IsDefined(typeof(Neo4jIgnoreAttribute))``` instead of ```pi.IsDefined(typeof(JsonIgnoreAttribute))``` ... should be a manageable risk.

As I'm not familiar with your testing engine, I don't know how to prove. Mayby you can use the sample code in #336 to test it by your own.

I just added you as a contributor to my branch.

PaddiM8 commented:
Any progress on this? This issue is preventing me from switching to bolt!

cskardon commented:
No, there has been no progress, unfortunately I have work and other commitments.

Let me do some explaining - in an attempt to not just come across as being someone who wants to block this and make your life hard for the sake of it.

By taking a PR on  - the owner of the repo is in effect saying that they are happy to take on the technical debt of a PR - i.e. it's not just _you_ that use this OSS project, but lots of other people. By saying something to the effect of:

>Works on my machine :tm:

...there is no evidence for me, nor anyone else who uses the project - and one of the big issues about that is that you might use little work-arounds that you have forgotten about in the past to achieve things - and when I publish this to nuget I suddenly get a lot of unhappy people complaining that their code no longer does what it used to.

Largely - you don't (and shouldn't) care - as once it's published - your code is fine. But for me - if there are issues - that's then something I _have_ to fix - which means I need to be confident that if I change something - I have tests to help me know that my fixes don't break other things. 

This is the reason why I have the stance about tests that I do.

And it's not just me - lots of other OSS projects have this stance, and normally - a lot of internal projects. I'm not trying to block the PR - genuinely I'd be happy to pull it in _if_ I could be sure that the code was entirely benign and safe and doesn't have side-effects that break the other users. 

If you think it's unreasonable to expect tests for code being put into a project, I'm not sure what else I can do to convince you. I think if you've used anything like NUnit, MSTest etc that XUnit really isn't a big step and if you can code the `Neo4jIgnoreAttribute` I would hope you could work out how to put `Assert.True` instead of `Assert.IsTrue` in your code.

Clooney24 commented:
Hi Chris,

I absolutely understand insisting to proper tests. I just implemented 2 test methods to check the behaviour of the new Neo4jIgnoreAttribute in https and bolt. I’m not sure if this is exactly what you expected and if I put them in the correct places.

For http I added the test method JsonSerializerShouldNotSerializeNeo4jIgnoreAttribute in UserSuppliedSerializationTests

For bolt I added the test method JsonSerializerShouldNotSerializeNeo4JIgnoreAttribute in BoltGraphClientTests

Regardless of my changes, I would like to mention that renaming a property using Json-Rename ([JsonProperty("NewName")]) doesn`t work in bolt but in https, so simply switching to bolt could result in problems if someone is using that. Maybe it might be a good idea to consider implementing a Neo4jPropertyAttribute that handles this identically in both implementations and is not depending on JsonProperty so – similar to my Neo4jIgnoreAttribute – behaviour can be exactly defined for Neo4j serialization and maybe different for other serialization needs.

I also noticed that the namespace of BoltGraphClientTests is Neo4jClient.Test.Extensions, I expected Neo4jClient.Test.BoltGraphClientTests – didn’t change it.

- Reiner

cskardon commented on Neo4jClient.Shared/Attributes/Neo4jIgnoreAttribute.cs:
```diff
@@ -0,0 +1,9 @@
+namespace Neo4jClient
+
+{
+    public class Neo4jIgnoreAttribute : Attribute
```
Could you add:
`[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]`

To this, just to restrict the usage

Clooney24 replied:
Done.

cskardon commented:
Hey @Clooney24 et al. 

I'm very sorry for the time this is taking - (has taken). 
I've put this into the `4.x` version of the codebase, not the 3.x one, which will be `4.0.0.15-prerelease` - At the moment, I think there will be maybe another 1 or 2 releases before removing the pre-release tag - we're pretty much at RC atm.

I'm closing this for this reason.

Thank you for doing the changes, and, again, sorry for the time
